### PR TITLE
IMPRO-888: Modifications to tests to avoid blending time coordinates

### PR DESCRIPTION
**Pull request by bayliffe**
This PR modifies the unit tests associated with weighted blending to ensure we are blending forecast reference time rather than time. The cubes used in testing now all have a single validity time for data across the blending coordinate.

I have made the test in weightedblendacrosswholedimension more rigorous as well. Any attempt to blend data at different times will raise an exception unless an exemption is provided when the instance of the plugin is created. This exemption has been added into the call from blend across adjacent points to allow for triangular time blending.

Needs to be rebased when IMPRO-888 has been merged.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py [hunk 1]
```diff
@@ -172,18 +174,20 @@ def setUp(self):
...
             dim_coords_and_dims=[(threshold, 0)])
 
         self.weights3d = Cube(
             weights3d, long_name='weights',
-            dim_coords_and_dims=[(time_coord, 0), (lat_coord, 1),
+            dim_coords_and_dims=[(frt_coord, 0), (lat_coord, 1),
                                  (lon_coord, 2)],
-            aux_coords_and_dims=[(frt_coord, 0), (fp_coord, 0)])
+            aux_coords_and_dims=[(fp_coord, 0)])
+        self.weights3d.add_aux_coord(time_coord)
```
Comment (cgsandford) [inline lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py]: Can use: `aux_coords_and_dims=[(fp_coord, 0), (time_coord, None)]` to get the same effect.

File: lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py [hunk 2]
```diff
@@ -232,16 +236,29 @@ class Test_check_compatible_time_points(Test_weighted_blend):
...
-    def test_forecast_reference_time_exception(self):
-        """Test that a ValueError is raised if the coordinate to be blended
-        is forecast_reference_time and the points on the time coordinate are
-        not equal."""
+    def test_unmatched_validity_time_exception(self):
+        """Test that a ValueError is raised if the validity time of the
+        slices over the blending coordinate differ. We should only be blending
+        data at equivalent times unless we are triangular time blending."""
+        self.cube.remove_coord('time')
+        self.cube.coord('forecast_reference_time').rename('time')
```
Comment (cgsandford) [inline lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py]: Would be good to update the error message below / above (in the code) to reflect the new logic.

File: lib/improver/blending/weighted_blend.py [hunk 3]
```diff
@@ -515,14 +527,16 @@ def check_compatible_time_points(self, cube):
...
-                raise ValueError(msg)
+        if self.timeblending is True:
+            return
+
+        time_points = cube.coord("time").points
+        if len(set(time_points)) > 1:
+            msg = ("For blending using the forecast_reference_time "
+                   "coordinate, the points within the time coordinate need"
+                   " to be the same. The time points within the input cube"
+                   " are {}".format(time_points))
```
Comment (cgsandford) [inline lib/improver/blending/weighted_blend.py]: Could we update this error message to reflect the new logic?

File: lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py (hunk 1 above)
Comment (bayliffe) [reply lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py]: Done.

File: lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py (hunk 2 above)
Comment (bayliffe) [reply lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py]: Done.

File: lib/improver/blending/weighted_blend.py (hunk 3 above)
Comment (bayliffe) [reply lib/improver/blending/weighted_blend.py]: Done.

File: lib/improver/blending/blend_across_adjacent_points.py [hunk 4]
```diff
@@ -103,7 +103,8 @@ def __init__(self, coord, central_point, parameter_units, width,
         # Set up the blending function, based on whether weighted blending or
         # maximum probabilities are needed.
         self.BlendingPlugin = (
-            WeightedBlendAcrossWholeDimension(coord, weighting_mode))
+            WeightedBlendAcrossWholeDimension(coord, weighting_mode,
+                                              timeblending=True))
```
Comment (markysparks) [inline lib/improver/blending/blend_across_adjacent_points.py]: Not part of your changes I know, but it seems there's an unused import at the top of this module:

`from improver.utilities.cube_manipulation import concatenate_cubes`

**Comment by bayliffe** [comment]
I have removed the unused import spotted by @markysparks. This was the only change requested, so if all the tests pass this can be merged.

**Pull request merged**